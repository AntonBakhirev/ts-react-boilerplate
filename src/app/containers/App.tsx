import {normalize, setupPage} from "csstips";
import * as React from "react";
import {Helmet} from "react-helmet";
import {addLocaleData, IntlProvider} from "react-intl";
import * as de from "react-intl/locale-data/de";
import * as en from "react-intl/locale-data/en";
import * as es from "react-intl/locale-data/es";
import * as fr from "react-intl/locale-data/fr";
import {connect} from "react-redux";
import {routeNodeSelector} from "redux-router5";
import {State as IRouteState} from "router5";
import {cssRaw, cssRule, style} from "typestyle";

import {Header} from "../components";
import {IStore} from "../redux/IStore";
import {IState} from "../redux/modules/baseModule";
import {ILanguage} from "../redux/modules/languageModule";
import {AboutPage} from "./AboutPage";
import {CounterPage} from "./CounterPage";
import {HomePage} from "./HomePage";
import {RegisterPage} from "./RegisterPage";
import {StarsPage} from "./StarsPage";

const appConfig = require("../../../config/main");

// Global style
cssRaw(`@import url(https://fonts.googleapis.com/css?family=Roboto);`);
normalize();
setupPage("#app");
cssRule(`html, body`, {
  fontFamily: "Roboto",
  height: "auto"
});

// App container style
const Styles = {
  container: style({
    margin: 0,
    padding: 0,
    textAlign: "center"
  })
};

class App extends React.Component<IStateToProps, null> {
  private components: any = {
    about: AboutPage,
    counter: CounterPage,
    home: HomePage,
    register: RegisterPage,
    stars: StarsPage
  };

  constructor() {
    super();
    addLocaleData([...en, ...es, ...fr, ...de]);
  }

  public render(): JSX.Element {
    const {language, route} = this.props;
    const segment = route ? route.name.split(".")[0] : undefined;
    return (
      <IntlProvider locale={language.payload.locale} messages={language.payload.languageData}>
        <section className={Styles.container}>
          <Helmet {...appConfig.app.head}/>
          <Header />
          {segment && this.components[segment] ? React.createElement(this.components[segment]) : <div>Not found</div>}
        </section>
      </IntlProvider>
    );
  }
}

interface IStateToProps {
  language: IState<ILanguage>;
  route: IRouteState;
}

const mapStateToProps = (state: Partial<IStore>) => ({
  languages: state.language,
  ...routeNodeSelector("")(state)
});

const connectedApp = connect<IStateToProps, null, null>(mapStateToProps, null)(App);

export {connectedApp as App, App as UnconnectedApp, Styles}
