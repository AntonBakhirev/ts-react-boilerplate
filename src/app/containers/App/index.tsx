const appConfig = require('../../../../config/main');

import * as React from 'react';
import * as Helmet from 'react-helmet';
import { style, cssRule } from 'typestyle';
import { SiteStyles } from '../../SiteStyles';
import { Header } from 'components';

// Define base stylings here
// Registering a CSS rule cannot use properties with nested styles

cssRule('html', {
    boxSizing: 'border-box',
});

cssRule('*, *:before, *:after', {
    boxSizing: 'inherit',
});

cssRule('body',
    SiteStyles.backgroundColor,
);

cssRule('a',
    SiteStyles.linkColor,
    { textDecoration: 'none' },
);

cssRule('a:hover',
    SiteStyles.linkColorOnHovered,
);

const Styles = {
    container: style({
        padding: 0,
        margin: 0,
        textAlign: 'center',
    }),
};

class App extends React.Component<any, any> {
  public render() {
    return (
      <section className={Styles.container}>
        <Helmet {...appConfig.app} {...appConfig.app.head}/>
        <Header />
        {this.props.children}
      </section>
    );
  }
}

export {Â App, Styles }
