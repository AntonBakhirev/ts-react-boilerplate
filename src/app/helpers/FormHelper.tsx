import * as React from "react";
import {FormattedMessage} from "react-intl";
import {connect} from "react-redux";
import {Field, GenericField, WrappedFieldInputProps, WrappedFieldMetaProps} from "redux-form";
import {style} from "typestyle";
import {IStore} from "../redux/IStore";

interface IFlexWrappedFieldProps {
  input?: Partial<WrappedFieldInputProps>;
  meta?: Partial<WrappedFieldMetaProps>;
}

// While waiting for this pull request https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17693,
// we overrides meta as any for now
interface IRenderFieldProps extends IFlexWrappedFieldProps {
  defaultMessage: string;
  languageId: string;
  type?: string;
}

export const FieldExtra = Field as new () => GenericField<IRenderFieldProps>;

export const required = (id, defaultMessage) => (value) => (value ? undefined : {id, defaultMessage});

export const maxLength = (id, defaultMessage) => (max) => (value) => value && value.length > max ? {id, defaultMessage, values: {max}} : undefined;

export const minLength = (id, defaultMessage) => (min) => (value) => value && value.length < min ? {id, defaultMessage, values: {min}} : undefined;

export const numberType = (id, defaultMessage) => (value) => value && isNaN(Number(value)) ? {id, defaultMessage} : undefined;

export const minValue = (id, defaultMessage) => (min) => (value) => value < min ? {id, defaultMessage, values: {min}} : undefined;

export const matchedPwd = (id, defaultMessage) => (value, otherValues) => value !== otherValues.password ? {id, defaultMessage} : undefined;

export const email = (id, defaultMessage) => (value) => value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value) ? {id, defaultMessage} : undefined;

export const tooOld = (id, defaultMessage) => (value) => value && value > 65 ? {id, defaultMessage} : undefined;

export const aol = (id, defaultMessage) => (value) => value && /.+@aol\.com/.test(value) ? {id, defaultMessage} : undefined;

/*tslint:disable:object-literal-sort-keys*/
const styles = {
  row: style({
    margin: "15px 0"
  }),
  inputItem: style({
    padding: 5,
    border: "1px solid #ccc"
  }),
  message: style({
    marginTop: 5
  }),
  error: style({
    color: "red"
  }),
  warning: style({
    color: "darkorange"
  })
};
/*tslint:enable:object-literal-sort-keys*/

/*tslint:disable:jsx-no-multiline-js*/
class CustomField extends React.Component<IStateToProps & IRenderFieldProps, null> {
  public render(): JSX.Element {
    const {defaultMessage, input, languageId, type, meta: {active, touched, error, warning}} = this.props;
    return (
      <div className={styles.row}>
        <FormattedMessage id={languageId} defaultMessage={defaultMessage} />
        <div>
          <input {...input} placeholder={this.props.languageData[languageId] || defaultMessage} type={type} className={styles.inputItem} />
          {
            (active || touched) &&
            (
              (
                error &&
                <div className={styles.message}>
                <span className={styles.error}>
                  <FormattedMessage id={error.id} defaultMessage={error.defaultMessage} values={error.values} />
                </span>
                </div>
              )
              ||
              (
                warning &&
                <div className={styles.message}>
                <span className={styles.warning}>
                  <FormattedMessage id={warning.id} defaultMessage={warning.defaultMessage} values={warning.values} />
                </span>
                </div>
              )
            )
          }
        </div>
      </div>
    );
  }
}

interface IStateToProps {
  languageData: any;
}

const mapStateToProps = (state: Partial<IStore>) => ({
  languageData: state.language.payload.languageData
});

const ConnectedCustomField = connect<IStateToProps, null, IRenderFieldProps>(mapStateToProps, null)(CustomField);

export {CustomField as UnconnectedCustomField, ConnectedCustomField as CustomField, mapStateToProps};
