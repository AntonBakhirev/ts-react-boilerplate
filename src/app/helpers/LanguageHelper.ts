import * as fs from "fs";

export class LanguageHelper {
  private preferedLang: string;
  constructor(private requestLanguage: string) {
    this.preferedLang = this.requestLanguage.split(",")[0];
  }

  public static getDefaultLanguage(): string {
    return "en-GB";
  }

  public static isSupported(language: string): boolean {
    return LanguageHelper.getSupportedLanguagesArray().indexOf(language) !== -1;
  }

  private static getLanguageFileLocation(language: string): string {
    return `${__dirname}/../translations/${language.toLowerCase()}.json`;
  }

  private static getSupportedLanguagesArray(): string[] {
    return ["en-GB", "en-US", "de"];
  }

  private static getLanguageData(language: string): object {
    return JSON.parse(fs.readFileSync(LanguageHelper.getLanguageFileLocation(language)).toString());
  }

  public getRequestLanguageData(): any {
    return LanguageHelper.getLanguageData(LanguageHelper.isSupported(this.preferedLang) ? this.preferedLang : "en-GB");
  }

  public getRequestLang(): string {
    return this.requestLanguage;
  }

  public getPreferredLanguage(): string {
    return this.preferedLang;
  }
}
