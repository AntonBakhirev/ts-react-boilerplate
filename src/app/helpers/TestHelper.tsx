/** React Specific */
import {mount, ReactWrapper} from "enzyme";
import * as React from "react";
import {IntlProvider} from "react-intl";
import {Provider} from "react-redux";
import {createStore} from "redux";
import rootReducer from "../redux/rootReducer";

/** Redux Mock Store Configuration */
import {IStore} from "../redux/IStore";

/** Render Component */
function renderComponent(ComponentClass: React.ComponentClass<any> | React.SFC<any> | React.ClassType<any, any, any> | string, state?: object, props?: object): ReactWrapper<IStore, any> {
  const store = createStore(rootReducer, state);

  return mount(
    <IntlProvider locale="en">
      <Provider store={store}>
        <ComponentClass {...props} />
      </Provider>
    </IntlProvider>
  );
}

export {renderComponent};
