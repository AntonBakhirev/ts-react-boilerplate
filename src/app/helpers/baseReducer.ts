import IBaseAction from "../models/IBaseAction";

/**
 * @param name todo: change to a better name
 * @param state
 * @param action
 * @returns {any}
 */
export default function baseReducer<TState, TAction>(name: string, state: TState, action: IBaseAction & TAction): TState {
  switch (action.type) {
    case name + "_PENDING":
      return Object.assign({}, state, {
        isFetching: true
      });

    case name + "_FULFILLED":
      return Object.assign({}, state, {
        isFetching: false,
        payload: action.payload
      });

    case name + "_REJECTED":
      return Object.assign({}, state, {
        error: true,
        isFetching: false,
        payload: action.payload
      });

    default:
      return state;
  }
}