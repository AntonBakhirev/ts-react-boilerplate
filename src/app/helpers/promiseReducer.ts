import {IAction, IState} from "../redux/modules/baseModule";

export interface IPromiseActions {
  FULFILLED: string;
  PENDING: string;
  REJECTED: string;
}

export const promiseAction = (baseAction: string): IPromiseActions => {
  return {
    FULFILLED: baseAction + "_FULFILLED",
    PENDING: baseAction + "_PENDING",
    REJECTED: baseAction + "_REJECTED"
  };
};

// tslint:disable:max-line-length
export default function promiseReducer<P, M = {}>(baseAction: string, state: IState<P, M>, action: IAction<P, M>): IState<P, M> {
  switch (action.type) {
    case promiseAction(baseAction).PENDING:
      return {
        ...state,
        isFetching: true
      };

    case promiseAction(baseAction).FULFILLED:
      return {
        ...state,
        isFetching: false,
        payload: action.payload
      };
    case promiseAction(baseAction).REJECTED:
      return {
        ...state,
        error: true,
        isFetching: false,
        message: action.message
      };
    default:
      return state;
  }
}
