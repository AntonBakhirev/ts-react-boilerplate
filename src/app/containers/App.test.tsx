import { chai, renderComponent } from "../helpers/TestHelper";
import { App, Styles } from "./App";

describe("<App />", () => {

  const component = renderComponent(App);

  it("Renders with correct style", () => {
    chai.expect(component).to.have.className(Styles.container);
  });

});
