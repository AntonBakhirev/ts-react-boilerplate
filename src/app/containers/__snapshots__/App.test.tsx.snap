// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<IntlProvider
  locale="en-GB"
  messages={
    Object {
      "greeting": "Hello!",
    }
  }
>
  <section
    className="f1whcvza"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
      meta={
        Array [
          Object {
            "charset": "utf-8",
          },
          Object {
            "content": "ie=edge",
            "http-equiv": "x-ua-compatible",
          },
          Object {
            "content": "width=device-width, initial-scale=1",
            "name": "viewport",
          },
          Object {
            "content": "React Redux Typescript",
            "name": "description",
          },
        ]
      }
      title="Crazy Factory"
    />
    <Header />
    <HomePage />
  </section>
</IntlProvider>
`;
